pykafka/rdkafka/_rd_kafkamodule.c

```diff
@@ -0,0 +1,1182 @@
+    Py_XDECREF(err_args);
+}
+
+
+/**
+ * Message type
+ */
+
+
+PyObject *Message;  /* pykafka.protocol.Message */
+
+
+/**
+ * Shared bits of Producer and Consumer types
+ */
+
+
+/* Note that with this RdkHandle, we hold a separate rd_kafka_t handle for each
+ * rd_kafka_topic_t, whereas librdkafka would allow sharing the same rd_kafka_t
+ * handle between many topic handles, which would be far more efficient.  The
+ * problem with that is that it would require the same rd_kafka_conf_t settings
+ * across all class instances sharing a handle, which is somewhat incompatible
+ * with the current pykafka API.
+ *
+ * We need a pthread rwlock here, because in many methods we release the GIL
+ * (this for various reasons - one key reason is that it prevents us
+ * deadlocking if we do blocking calls into librdkafka and then get callbacks
+ * out of librdkafka, and the callbacks would try to grab the GIL).  Once we
+ * release the GIL however, there may be other threads calling RdkHandle_stop
+ * (which in pykafka can happen on any thread).  The rule here then is that
+ * RdkHandle_stop needs to take out an exclusive lock (wrlock), whereas most
+ * other calls are safe when taking out a shared lock (rdlock). */
+typedef struct {
+    PyObject_HEAD
+    pthread_rwlock_t rwlock;
+    rd_kafka_t *rdk_handle;
+    rd_kafka_conf_t *rdk_conf;
+    rd_kafka_topic_t *rdk_topic_handle;
+    rd_kafka_topic_conf_t *rdk_topic_conf;
+
+    /* Consumer-specific fields */
+    rd_kafka_queue_t *rdk_queue_handle;
+    PyObject *partition_ids;
+
+    /* Producer-specific fields; see Producer_produce for details */
+    PyObject *pending;
+} RdkHandle;
+
+
+/* Only for inspection; if you'd manipulate these from outside the module, we
+ * have no error checking in place to protect from ensuing mayhem */
+static PyMemberDef RdkHandle_members[] = {
+    {"_partition_ids", T_OBJECT_EX, offsetof(RdkHandle, partition_ids),
+                       READONLY, "Partitions fetched from by this consumer"},
+    {"_pending_messages", T_OBJECT_EX, offsetof(RdkHandle, pending),
+                          READONLY, "Messages pending a delivery report"},
+    {NULL}
+};
+
+
+static PyObject *
+RdkHandle_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *self = PyType_GenericNew(type, args, kwds);
+    if (self) {
+        int res = pthread_rwlock_init(&((RdkHandle *)self)->rwlock, NULL);
+        if (res) {
+            Py_DECREF(self);
+            return set_pykafka_error("RdKafkaException", "Failed rwlock init");
+        }
+    }
+    return self;
+}
+
+
+static int
+RdkHandle_unlock(RdkHandle *self)
+{
+    if (pthread_rwlock_unlock(&self->rwlock)) {
+        set_pykafka_error("RdKafkaException", "Failed to release rwlock");
+        return -1;
+    }
+    return 0;
+}
+
+
+/* Get shared lock and optionally check handle is running.  Returns non-zero
+ * if error has been set */
+static int
+RdkHandle_safe_lock(RdkHandle *self, int check_running)
+{
```

Comment:
I think all (or at least many) of the functions in this file would benefit from docstrings indicating how they fit into the overall operation of the producer/consumer.